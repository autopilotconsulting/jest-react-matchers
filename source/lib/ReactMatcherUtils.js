import { propertiesForElement } from 'describe-react-element';
import flatten from 'array-flatten';

const ReactMatcherUtils = {
  elementsMatch: function(equals, actualElement, expected) {
    if (actualElement === null || typeof(actualElement) === 'undefined') { return false; }
    if (typeof(actualElement) === 'string') { return expected === actualElement; }
    if (expected.type !== actualElement.type) { return false; }

    const { children: expectedChildren, ...expectedProps } = propertiesForElement(expected);
    const { children: actualChildren, ...actualProps } = propertiesForElement(actualElement);

    const propertiesMatcher = jasmine.objectContaining(expectedProps);
    const propertiesMatch = equals(actualProps, propertiesMatcher);

    if (!propertiesMatch) { return false; }

    const expectedChildrenMatch = this.collectionHasMatchForAllElements(equals, actualChildren, expectedChildren);
    return expectedChildrenMatch;
  },

  collectionHasMatchForElement: function(equals, actualCollection, expected) {
    const test = (actualElement) => this.elementsMatch(equals, actualElement, expected);
    return actualCollection.some(test);
  },

  collectionHasMatchForAllElements: function(equals, actualCollection, expectedCollection) {
    if(typeof(expectedCollection) === 'undefined') { return true; }

    expectedCollection = flatten([expectedCollection]);
    actualCollection = flatten([actualCollection]);

    const test = (expectedElement) => this.collectionHasMatchForElement(equals, actualCollection, expectedElement);
    return expectedCollection.every(test);
  },

  findMatchingElement: function(equals, actualElement, expectedElement) {
    const matches = [];

    if(typeof(actualElement) === 'undefined') { return; }

    if(this.elementsMatch(equals, actualElement, expectedElement)) {
      matches.push(actualElement);
    }

    if (actualElement.props) {
      const children = flatten([actualElement.props.children]);
      const matchesFromChildren = children.map((child) => {
        return this.findMatchingElement(equals, child, expectedElement);
      });

      const matchCollection = flatten(matchesFromChildren).filter(element => element);
      matches.push(...matchCollection);
    }

    if(matches.length === 0) { return; }
    return matches.length === 1 ? matches[0] : matches;
  }
};

export default ReactMatcherUtils;