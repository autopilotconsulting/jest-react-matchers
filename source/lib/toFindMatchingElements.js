import { describeReactElement } from 'describe-react-element';
import ReactMatcherUtils from './ReactMatcherUtils';
import groupArray from './groupArray';

import flatten from 'array-flatten';

function toFindMatchingElements(actualElement, ...expectedElements) {
  const actualDescription = describeReactElement(actualElement, 1);

  expectedElements = flatten([expectedElements]);
  const { true: hits, false: misses } = groupArray(expectedElements, (expectedElement) => {
    let matches = ReactMatcherUtils.findMatchingElement(this.equals, actualElement, expectedElement);
    matches = flatten([matches]).filter(element => element);
    return matches.length > 0;
  });

  const isCorrect = this.isNot ? !hits : !misses;
  const result = { pass: isCorrect != this.isNot };

  if (isCorrect) { return result; }

  const describeExpected = (element) => describeReactElement(element, 1);
  if (this.isNot) {
    const hitDescriptions = hits.map(describeExpected);

    result.message = () => [
      'The following unexpected elements were found:',
      ...hitDescriptions,
      '',
      'in:',
      actualDescription,
    ].join("\n");
  } else {
    const missDescriptions = misses.map(describeExpected);

    result.message = () => [
      'The following elements were not found:',
      ...missDescriptions,
      '',
      'in:',
      actualDescription,
    ].join("\n");
  }

  return result;
};

export default toFindMatchingElements;