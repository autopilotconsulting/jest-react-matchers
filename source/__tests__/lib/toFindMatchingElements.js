import React from 'react';
import { equals } from '../../../support/jasmine_utils';

import ReactMatcherUtils from '../../lib/ReactMatcherUtils.js';
import toFindMatchingElements from '../../lib/toFindMatchingElements.js';

describe('ReactMatchers', () => {
  describe('#toFindMatchingElements', () => {
    let matcherContext;
    
    const actual = (
      <div>
        <div>
          <span>Pippy</span>
        </div>
        <li>Hatch</li>
        <div>
          <p>
            <li data-tis-wife={true}>Lu</li>
          </p>
        </div>
      </div>
    );

    beforeEach(() => {
      matcherContext = { equals, isNot: false };
    });

    it('should pass if all expected elements are present', () => {
      let result = toFindMatchingElements.call(matcherContext, actual, <li />);
      expect(result.pass).toBe(true);

      result = toFindMatchingElements.call(matcherContext, actual, [<span>Pippy</span>, <li>Hatch</li>]);
      expect(result.pass).toBe(true);

      result = toFindMatchingElements.call(matcherContext, actual, <span>Pippy</span>, <li>Hatch</li>);
      expect(result.pass).toBe(true);
    });

    it('should not pass if any expected elements are missing', () => {
      let result = toFindMatchingElements.call(matcherContext, actual, [<span>Pippy</span>, <marquee />]);
      expect(result.pass).toBe(false);
    });

    describe('when negated', () => {
      it('should "pass" if any unexpected elements are present', () => {
        matcherContext.isNot = true;
        let result = toFindMatchingElements.call(matcherContext, actual, [<span>Pippy</span>, <marquee />]);
        expect(result.pass).toBe(true);
      });

      it('should not "pass" if all unexpected elements are absent', () => {
        matcherContext.isNot = true;
        let result = toFindMatchingElements.call(matcherContext, actual, [<table />, <marquee />]);
        expect(result.pass).toBe(false);
      });
    });

    it('should describe a failure appropriately', () => {
      const expected = [
        <li data-tis-wife={true} />,
        <li>Lu</li>,
        'Pippy',
        <div data-awesome={true}>An awesome div!!</div>,
        <span>Lame Span</span>
      ];

      const result = toFindMatchingElements.call(matcherContext, actual, expected);      
      expect(result.message).toMatchSnapshot();
    });

    it('should describe a negated failure appropriately', () => {
      const unexpected = [
        <li data-tis-wife={true} />,
        <li>Lu</li>,
        'Pippy',
        <div data-awesome={true}>An awesome div!!</div>,
        <span>Lame Span</span>
      ];

      matcherContext.isNot = true;
      const result = toFindMatchingElements.call(matcherContext, actual, unexpected);
      expect(result.message).toMatchSnapshot();
    });
  });
});