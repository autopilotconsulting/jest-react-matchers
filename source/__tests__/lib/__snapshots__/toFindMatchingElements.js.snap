// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactMatchers #toFindMatchingElements should describe a failure appropriately 1`] = `
"The following elements were not found:
  <div data-awesome='true'>
    An awesome div!!
  </div>
  <span>
    Lame Span
  </span>

in:
  <div>
    <div>
      <span>
        Pippy
      </span>
    </div>
    <li>
      Hatch
    </li>
    <div>
      <p>
        <li data-tis-wife='true'>
          Lu
        </li>
      </p>
    </div>
  </div>"
`;

exports[`ReactMatchers #toFindMatchingElements should describe a negated failure appropriately 1`] = `
"The following unexpected elements were found:
  <li data-tis-wife='true' />
  <li>
    Lu
  </li>
  Pippy

in:
  <div>
    <div>
      <span>
        Pippy
      </span>
    </div>
    <li>
      Hatch
    </li>
    <div>
      <p>
        <li data-tis-wife='true'>
          Lu
        </li>
      </p>
    </div>
  </div>"
`;
